### ISSUE TITLE
Happy Eyeballs?

### ISSUE THREAD

**Issue by DavidSchinazi** [ietf-wg-masque/draft-ietf-masque-quic-proxy#59]
CONNECT-UDP can't perform Happy Eyeballs because UDP doesn't have an equivalent of the TCP SYN-ACK. However, QUIC has such an equivalent. This extension could recommend performing Happy Eyeballs when target_host is a DNS host name.

**Comment by gloinul** [comment]
Have to agree with @LPardue here. I don't see how this can work. A UDP Proxy can clearly run happy eyeball and race the quick connection between the masque client and the masque server (proxy) over IPv4 and IPv6 and in this case it would be the MASQUE used HTTP establishment that is raced. For the end-to-end connection that is being tunnelled, that require the application above the QUIC connection(s) to perform happy eyeballs. If you want to enable this, then you need to have the end-to-end application open two QUIC connection over the same proxy and race these and thus indicate a preference for IPv4 or IPv6 when doing the request for proxying. 
 
 I don't believe in the proxy duplicating what it thinks is a QUIC handshake packet for the E2E quic connection and send it over both IPv4 and IPv6 for domains that have both A and AAAA DNS records. If you one would do that and then select which to use based on the response. Then first you are impacting when the response can be sent. So any attempt in this direction would need an extension and explicit signalling about it.

**Comment by tfpauly** [comment]
Yeah, the client would need to be involved in order to run proper happy eyeballs, I think.

The proxy could certainly learn for commonly accessed targets that a particular address actually doesn't work — like if the DNS entry had IPv6 addresses, but never responded, for future requests the proxy could prefer IPv6.

**Comment by LPardue** [comment]
Proxy heuristics seem fine but then that seems common to any type of HTTP CONNECT proxy. Is there anything special to say here?

**Comment by DavidSchinazi** [comment]
(For the sake of simplicity, I'm assuming there's only one v4 address and one v6 address below, but this generalizes)

My initial thought was to have the proxy send the INITIAL to both the v4 and v6 target addresses (staggered as in HEv2, not all at once) and then pick the first target address that replies as its target, and forward the reply to the client. If the proxy receives a reply from the other address, there are two scenarios:
1. the second reply has packet number 0: this means that it came from a different target server - the proxy drops the reply and generates a fake INITIAL CONNECTION_CLOSE and sends it to the second target address to tell that server to give up
2. the second reply has packet number > 0: this means that it was the same target server as the first reply - the proxy forwards the reply to the client and picks the best target address as its future target address to send to

Of course we'd need something a bit more smart to handle packet loss but I think the general idea would work.

If folks think this is too complicated I'm happy to take that to an extension, but since we haven't adopted this yet I thought it might be worth thinking about since we're not delaying progress.

**Comment by LPardue** [comment]
My concerns stem from implications of reuse of connection IDs and tokens, e.g.

> An endpoint MUST NOT reuse a connection ID when sending from more than one local address -- for example, when initiating connection migration

Maybe we could make this more of explicit behaviour? E.g. the client gives a proxy multiple SCIDS to use for this and then just returns packets with the selected one

**Comment by DavidSchinazi** [comment]
That MUST NOT is post-handshake though. The client not allowed to switch its SCID during the handshake (except to the one the server gives it)

**Comment by LPardue** [comment]
Yeah agree it's a tradeoff. 

Thinking aloud.
It seems like in an era of HTTPS records, happy eyeballs for QUIC could be useful for non-proxy clients. And guidance on how to close an uneeded connection seems useful. The decision whether to duplicate Initials could be part of guidance around tradeoffs. A proxy could then take that and use it. A CONNECT-UDP proxy that snoops could even employ the approach ypu outline

**Comment by tfpauly** [comment]
Discussed with David, OK to close

**Comment by DavidSchinazi** [comment]
To add context from that offline discussion here: both Tommy's and my deployment end up only using IPv6 to the target, so Happy Eyeballs isn't a pressing need for us. Since this can be added in a future document just as easily, it makes sense to punt it for now.
